spec/controllers/api/groups_controller_spec.rb

```diff
@@ -270,6 +270,109 @@
         end
       end
     end
+    context 'POST add_extra_marks' do
+      let(:submission) { create(:version_used_submission, grouping: grouping) }
+      context 'add extra_mark' do
+        let(:old_mark) { submission.get_latest_result.total_mark }
+        before :each do
+          old_mark
+          post :create_extra_marks, params: { assignment_id: grouping.assignment.id,
+                                              id: grouping.group.id,
+                                              extra_marks: 10.0,
+                                              description: 'sample' }
+          grouping.reload
+        end
+        it 'should add new extra mark' do
+          result = submission.get_latest_result
+          added_extra_mark = result.extra_marks.last
+          expect(added_extra_mark.extra_mark).to eq(10.0)
+        end
+        it 'should update total_mark' do
+          result = submission.get_latest_result
+          new_total_mark = result.total_mark
+          expect(old_mark + 10.0).to eq(new_total_mark)
+        end
+        it 'should respond with 200' do
+          expect(response.status).to eq(200)
+        end
+      end
+      context 'add wrong extra_mark' do
+        let(:old_mark) { submission.get_latest_result.total_mark }
+        before :each do
+          old_mark
+          post :create_extra_marks, params: { assignment_id: grouping.assignment.id,
+                                              id: grouping.group.id,
+                                              extra_marks: 'a',
+                                              description: 'sample' }
+          grouping.reload
+        end
+        it 'should respond with 500' do
+          expect(response.status).to eq(500)
+        end
+        it 'should not update the total mark' do
+          result = submission.get_latest_result
+          new_total_mark = result.total_mark
+          expect(old_mark).to eq(new_total_mark)
+        end
+      end
+      describe 'when the arguments are invalid' do
+        it_behaves_like 'Invalid_arguments'
+      end
+    end
+    context 'DELETE remove_extra_marks' do
+      describe 'when the arguments are invalid' do
+        it_behaves_like 'Invalid_arguments'
+      end
+      describe 'when the arguments are valid' do
+        let(:submission) { create(:version_used_submission, grouping: grouping) }
+        before :each do
+          submission
+          post :create_extra_marks, params: { assignment_id: grouping.assignment.id,
```

Comment:
please create the extra mark before the tests using factories. If you create a mark by calling this route and there is a bug in the `create_extra_marks` method, these tests will fail for the wrong reasons.